ByteBuddy-MockMaker uses wrong ClassLoader on Java 21, if mocked type and additional interfaces are from different classloaders

### Describe the bug

Given the following conditions:

- ByteBuddy is available
- We are running on Java 21
- and the following classloader hierarchy:

![Unbenannt-2024-07-14-2000](https://github.com/user-attachments/assets/5370dce8-aff8-4f3d-bdf1-92214799d7d9)

When mocking type `iam.Mocked` with `additionalInterfaces: [foo.Bar]` and if the specification is loaded by `MySpecClassLoader`, we will get the following exception:

<details>
<summary>Stacktrace:</summary>

```
java.lang.IllegalArgumentException: Could not create type
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:170)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:399)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:190)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:410)
	at org.spockframework.mock.runtime.ByteBuddyMockFactory.createMock(ByteBuddyMockFactory.java:97)
	at org.spockframework.mock.runtime.ByteBuddyMockMaker.makeMock(ByteBuddyMockMaker.java:63)
	at org.spockframework.mock.runtime.MockMakerRegistry.createWithAppropriateMockMaker(MockMakerRegistry.java:194)
	at org.spockframework.mock.runtime.MockMakerRegistry.makeMockInternal(MockMakerRegistry.java:153)
	at org.spockframework.mock.runtime.MockMakerRegistry.makeMock(MockMakerRegistry.java:125)
	at org.spockframework.mock.runtime.JavaMockFactory.createInternal(JavaMockFactory.java:56)
	at org.spockframework.mock.runtime.JavaMockFactory.create(JavaMockFactory.java:42)
	at org.spockframework.mock.runtime.CompositeMockFactory.create(CompositeMockFactory.java:44)
	at org.spockframework.lang.SpecInternals.createMock(SpecInternals.java:55)
	at org.spockframework.lang.SpecInternals.createMockImpl(SpecInternals.java:330)
	at org.spockframework.lang.SpecInternals.createMockImpl(SpecInternals.java:320)
	at org.spockframework.lang.SpecInternals.MockImpl(SpecInternals.java:123)
	at apackage.ASpec.a feature(Script_2ea152aea2eb5fd1885101de83bd8e60.groovy:1)
Caused by: java.lang.NoClassDefFoundError: foo/Bar
	at java.base/java.lang.System$2.defineClass(System.java:2394)
	at net.bytebuddy.utility.dispatcher.JavaDispatcher$Dispatcher$ForNonStaticMethod.invoke(JavaDispatcher.java:1033)
	at net.bytebuddy.utility.dispatcher.JavaDispatcher$ProxiedInvocationHandler.invoke(JavaDispatcher.java:1163)
	at net.bytebuddy.dynamic.loading.ClassInjector$UsingLookup.injectRaw(ClassInjector.java:1639)
	at net.bytebuddy.dynamic.loading.ClassInjector$AbstractBase.inject(ClassInjector.java:119)
	at net.bytebuddy.dynamic.loading.ClassLoadingStrategy$UsingLookup.load(ClassLoadingStrategy.java:519)
	at net.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:101)
	at net.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:6367)
	at org.spockframework.mock.runtime.ByteBuddyMockFactory.lambda$createMock$2(ByteBuddyMockFactory.java:137)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:168)
	... 16 more
Caused by: java.lang.ClassNotFoundException: foo.Bar
	at java.base/java.lang.ClassLoader.findClass(ClassLoader.java:733)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 26 more
```

</details>

Spock will pick the ByteBuddy MockMaker and then use the `net.bytebuddy.dynamic.loading.ClassLoadingStrategy.UsingLookup` classloading strategy. This strategy receives the lookup of the to-be-mocked class. It then uses that lookup to define the mocked class. Internally, this just means that the classloader of the to-be-mocked is used. In our example, this would be `MockedClassClassLoader`. This classloader is not able to see `foo.Bar` and therefore loading the mocked class fails.

### To Reproduce

Paste the following testcase in `spock.util.EmbeddedSpecRunnerClassLoaderSpec`:

```groovy
  def "ByteBuddy Mocks with interfaces that are not visible to the mocked type's classloader"() {
    def testCl1 = new ByteBuddyTestClassLoader()
    def testCl2 = new ByteBuddyTestClassLoader()
    testCl1.defineInterface("foo.Bar")
    testCl2.defineClass("iam.Mocked")
    def cl = new MultipleParentClassLoader([getClass().classLoader, testCl1, testCl2])
    def runner = new EmbeddedSpecRunner(cl)
    when:
    def r = runner.runFeatureBody("""
def mock = Mock(iam.Mocked, additionalInterfaces: [foo.Bar])
expect:
true
""")
    then:
    noExceptionThrown()
  }
```

and run `gradlew :spock-specs:test --tests "spock.util.EmbeddedSpecRunnerClassLoaderSpec.ByteBuddy Mocks with interfaces that are not visible to the mocked type's classloader" "-Dvariant=4.0" "-DjavaVersion=21"`

### Expected behavior

The mocking succeeds

### Actual behavior

See exception above

### Java version

openjdk version "21.0.2" 2024-01-16

### Buildtool version

Gradle 8.10.2

### What operating system are you using

Windows

### Dependencies

Reproducer works in spock repo

### Additional context

Mockito has a similar problem. They only use the Lookup strategy for "local mocks" (see [here](https://github.com/mockito/mockito/blob/99426415c0ceb30e55216c3934854528c83f410e/mockito-core/src/main/java/org/mockito/internal/creation/bytebuddy/SubclassBytecodeGenerator.java#L165-L166)). Local mocks are mocks where the mocked type and all its additional interfaces are loadable by the classloader of the mocked type. I'd suggest implementing the same behavior for Spock and falling back to another loading strategy for non-local mocks. I have a simple version of this change ready and would love to open a PR if we can agree on this solution :)

Also FYI: I discovered this bug together with @AndreasTu

--- a/docs/release_notes.adoc
+++ b/docs/release_notes.adoc
@@ -20,6 +20,7 @@ include::include.adoc[]
 ** This enables loading of optional classes in e.g. OSGi environments
 * `EmbeddedSpecRunner` and `EmbeddedSpecCompiler` now support the construction with a custom `ClassLoader` spockPull:1988[]
 ** This allows the use of these classes in an OSGi environment, where the class imports in the embedded spec are not visible to the Spock OSGi bundle ClassLoader
+* Fix mocking issue with the ByteBuddy MockMaker when using multiple classloaders in Java 21 spockIssue:2017[]
 
 == 2.4-M4 (2024-03-21)
 

--- a/spock-core/src/main/java/org/spockframework/mock/runtime/ByteBuddyMockFactory.java
+++ b/spock-core/src/main/java/org/spockframework/mock/runtime/ByteBuddyMockFactory.java
@@ -84,13 +84,14 @@ class ByteBuddyMockFactory {
   }
 
   /**
-   * {@return true, if the mock is considered local}
+   * Returns {@code true}, if the mock is considered local
    * <p>
    * A mock is considered local, if all additional interfaces of the mock (including {@link ISpockMockObject}) are
    * loadable by the target class' classloader.
    *
    * @param targetClass The to-be-mocked class
    * @param additionalInterfaces Additional interfaces of the to-be-mocked type
+   * @return true, if this is a local mock. Otherwise false
    */
   @VisibleForTesting
   static boolean isLocalMock(Class<?> targetClass, Collection<Class<?>> additionalInterfaces) {

--- a/spock-core/src/main/java/org/spockframework/mock/runtime/ByteBuddyMockFactory.java
+++ b/spock-core/src/main/java/org/spockframework/mock/runtime/ByteBuddyMockFactory.java
@@ -28,6 +28,7 @@ import net.bytebuddy.description.modifier.Visibility;
 import net.bytebuddy.dynamic.Transformer;
 import net.bytebuddy.dynamic.loading.ClassInjector;
 import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
+import net.bytebuddy.dynamic.loading.MultipleParentClassLoader;
 import net.bytebuddy.dynamic.scaffold.TypeValidation;
 import net.bytebuddy.implementation.FieldAccessor;
 import net.bytebuddy.implementation.MethodDelegation;
@@ -37,10 +38,9 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.VisibleForTesting;
 import org.spockframework.mock.ISpockMockObject;
 import org.spockframework.mock.codegen.Target;
-import org.spockframework.util.Nullable;
 
 import java.lang.reflect.Method;
-import java.util.List;
+import java.util.Collection;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ThreadLocalRandom;
 
@@ -83,6 +83,26 @@ class ByteBuddyMockFactory {
     }
   }
 
+  /**
+   * {@return true, if the mock is considered local}
+   * <p>
+   * A mock is considered local, if all additional interfaces of the mock (including {@link ISpockMockObject}) are
+   * loadable by the target class' classloader.
+   *
+   * @param targetClass The to-be-mocked class
+   * @param additionalInterfaces Additional interfaces of the to-be-mocked type
+   */
+  @VisibleForTesting
+  static boolean isLocalMock(Class<?> targetClass, Collection<Class<?>> additionalInterfaces) {
+    // Inspired by https://github.com/mockito/mockito/blob/99426415c0ceb30e55216c3934854528c83f410e/mockito-core/src/main/java/org/mockito/internal/creation/bytebuddy/SubclassBytecodeGenerator.java#L165-L166
+    ClassLoader cl = new MultipleParentClassLoader.Builder()
+      .appendMostSpecific(targetClass)
+      .appendMostSpecific(additionalInterfaces)
+      .appendMostSpecific(ISpockMockObject.class)
+      .build();
+    return cl == targetClass.getClassLoader();
+  }
+
   Object createMock(IMockMaker.IMockCreationSettings settings) {
     final Class<?> type = settings.getMockType();
     TypeCache.SimpleKey key = new TypeCache.SimpleKey(type, settings.getAdditionalInterface());
@@ -99,7 +119,8 @@ class ByteBuddyMockFactory {
         }
         int randomNumber = Math.abs(ThreadLocalRandom.current().nextInt());
         String name = String.format("%s$%s$%d", typeName, "SpockMock", randomNumber);
-        ClassLoadingStrategy<ClassLoader> strategy = determineBestClassLoadingStrategy(targetClass);
+        ClassLoadingStrategy<ClassLoader> strategy = determineBestClassLoadingStrategy(targetClass, settings);
+        //noinspection resource
         return new ByteBuddy()
           .with(TypeValidation.DISABLED) // https://github.com/spockframework/spock/issues/776
           .ignore(none())
@@ -190,8 +211,8 @@ class ByteBuddyMockFactory {
   }
 
   @NotNull
-  private static ClassLoadingStrategy<ClassLoader> determineBestClassLoadingStrategy(Class<?> targetClass) throws Exception {
-    if (ClassInjector.UsingLookup.isAvailable()) {
+  private static ClassLoadingStrategy<ClassLoader> determineBestClassLoadingStrategy(Class<?> targetClass, IMockMaker.IMockCreationSettings settings) throws Exception {
+    if (ClassInjector.UsingLookup.isAvailable() && isLocalMock(targetClass, settings.getAdditionalInterface())) {
       Class<?> methodHandlesClass = Class.forName("java.lang.invoke.MethodHandles");
       Class<?> lookupClass = Class.forName("java.lang.invoke.MethodHandles$Lookup");
       Method lookupMethod = methodHandlesClass.getMethod("lookup");

--- a/spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockMakerSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockMakerSpec.groovy
@@ -18,6 +18,7 @@ package org.spockframework.mock.runtime
 
 import net.bytebuddy.dynamic.loading.MultipleParentClassLoader
 import org.spockframework.mock.CannotCreateMockException
+import spock.lang.Issue
 import spock.lang.Specification
 import spock.lang.Unroll
 import spock.mock.DetachedMockFactory
@@ -154,13 +155,16 @@ class ByteBuddyMockMakerSpec extends Specification {
     ex.message.startsWith("Cannot create mock for interface java.lang.Runnable. byte-buddy: The class AdditionalInterface is not visible by the classloader")
   }
 
+  @Issue("https://github.com/spockframework/spock/issues/2017")
   def "ByteBuddy Mocks with interfaces that are not visible to the mocked type's classloader"() {
+    given:
     def testCl1 = new ByteBuddyTestClassLoader()
     def testCl2 = new ByteBuddyTestClassLoader()
     testCl1.defineInterface("foo.Bar")
     testCl2.defineClass("iam.Mocked")
     def cl = new MultipleParentClassLoader([getClass().classLoader, testCl1, testCl2])
     def runner = new EmbeddedSpecRunner(cl)
+
     when: "A type is mocked with an additional interface from a classloader outside the mocked type's parent chain"
     runner.addClassImport(MockMakers)
     runner.runFeatureBody("""
@@ -168,6 +172,7 @@ def mock = Mock(iam.Mocked, mockMaker: MockMakers.byteBuddy, additionalInterface
 expect:
 true
 """)
+
     then:
     noExceptionThrown()
   }

--- a/spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockMakerSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyMockMakerSpec.groovy
@@ -16,10 +16,13 @@
 
 package org.spockframework.mock.runtime
 
+import net.bytebuddy.dynamic.loading.MultipleParentClassLoader
 import org.spockframework.mock.CannotCreateMockException
 import spock.lang.Specification
+import spock.lang.Unroll
 import spock.mock.DetachedMockFactory
 import spock.mock.MockMakers
+import spock.util.EmbeddedSpecRunner
 
 import java.lang.reflect.Proxy
 import java.util.concurrent.Callable
@@ -132,7 +135,7 @@ class ByteBuddyMockMakerSpec extends Specification {
 
     when:
     Mock(interfaceClass, mockMaker: MockMakers.byteBuddy)
-    
+
     then:
     CannotCreateMockException ex = thrown()
     ex.message.startsWith("Cannot create mock for interface Interface. byte-buddy: The class Interface is not visible by the classloader")
@@ -149,6 +152,36 @@ class ByteBuddyMockMakerSpec extends Specification {
     then:
     CannotCreateMockException ex = thrown()
     ex.message.startsWith("Cannot create mock for interface java.lang.Runnable. byte-buddy: The class AdditionalInterface is not visible by the classloader")
+  }
+
+  def "ByteBuddy Mocks with interfaces that are not visible to the mocked type's classloader"() {
+    def testCl1 = new ByteBuddyTestClassLoader()
+    def testCl2 = new ByteBuddyTestClassLoader()
+    testCl1.defineInterface("foo.Bar")
+    testCl2.defineClass("iam.Mocked")
+    def cl = new MultipleParentClassLoader([getClass().classLoader, testCl1, testCl2])
+    def runner = new EmbeddedSpecRunner(cl)
+    when: "A type is mocked with an additional interface from a classloader outside the mocked type's parent chain"
+    runner.addClassImport(MockMakers)
+    runner.runFeatureBody("""
+def mock = Mock(iam.Mocked, mockMaker: MockMakers.byteBuddy, additionalInterfaces: [foo.Bar])
+expect:
+true
+""")
+    then:
+    noExceptionThrown()
+  }
+
+  @Unroll("For target class #targetClass with additional interfaces #additionalInterfaces, isLocal() == #expected")
+  def "Local mocks are detected correctly"(Class<?> targetClass, Collection<Class<?>> additionalInterfaces, boolean expected) {
+    expect:
+    ByteBuddyMockFactory.isLocalMock(targetClass, additionalInterfaces) == expected
+
+    where:
+    targetClass | additionalInterfaces                                    | expected
+    DataClass   | []                                                      | true
+    DataClass   | [Serializable]                                          | true
+    DataClass   | [new ByteBuddyTestClassLoader().defineInterface("foo")] | false
   }
 }
 

--- a/spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyTestClassLoader.java
+++ b/spock-specs/src/test/groovy/org/spockframework/mock/runtime/ByteBuddyTestClassLoader.java
@@ -50,7 +50,7 @@ public final class ByteBuddyTestClassLoader extends ClassLoader {
   }
 
   /**
-   * Defines an empty interface with the passed {@code node}.
+   * Defines an empty interface with the passed {@code name}.
    *
    * @param name the name of the interface
    * @return the loaded {@code Class}
@@ -59,6 +59,22 @@ public final class ByteBuddyTestClassLoader extends ClassLoader {
     //noinspection resource
     return cache.computeIfAbsent(name, nameKey -> new ByteBuddy()
       .makeInterface()
+      .name(nameKey)
+      .make()
+      .load(this, loadingStrategy)
+      .getLoaded());
+  }
+
+  /**
+   * Defines an empty class with the passed {@code name}
+   *
+   * @param name the name of the class
+   * @return the loaded {@link  Class}
+   */
+  public synchronized Class<?> defineClass(String name) {
+    //noinspection resource
+    return cache.computeIfAbsent(name, nameKey -> new ByteBuddy()
+      .subclass(Object.class)
       .name(nameKey)
       .make()
       .load(this, loadingStrategy)

